pcsx2/IPC.cpp

```diff
@@ -0,0 +1,378 @@
+		return;
+	}
+#endif
+
+	// maximum queue of 4096 commands before refusing, approximated to the
+	// nearest legal value. We do not use SOMAXCONN as windows have this idea
+	// that a "reasonable" value is 5, which is not.
+	listen(m_sock, 4096);
+
+	// we save a handle of the main vm object
+	m_vm = vm;
+
+	// we start the thread
+	Start();
+}
+
+void SocketIPC::ExecuteTaskInThread()
+{
+	m_end = false;
+
+	// we allocate once buffers to not have to do mallocs for each IPC
+	// request, as malloc is expansive when we optimize for µs.
+	m_ret_buffer = new char[MAX_IPC_RETURN_SIZE];
+	m_ipc_buffer = new char[MAX_IPC_SIZE];
+	while (true)
+	{
+		m_msgsock = accept(m_sock, 0, 0);
+		if (m_msgsock == -1)
+		{
+			// everything else is non recoverable in our scope
+			// we also mark as recoverable socket errors where it would block a
+			// non blocking socket, even though our socket is blocking, in case
+			// we ever have to implement a non blocking socket.
+#ifdef _WIN32
+			int errno_w = WSAGetLastError();
+			if (!(errno_w == WSAECONNRESET || errno_w == WSAEINTR || errno_w == WSAEINPROGRESS || errno_w == WSAEMFILE || errno_w == WSAEWOULDBLOCK))
+			{
+#else
+			if (!(errno == ECONNABORTED || errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK))
+			{
+#endif
+				fprintf(stderr, "IPC: An unrecoverable error happened! Shutting down...\n");
+				m_end = true;
+				break;
+			}
+		}
+		else
+		{
+
+#ifdef _WIN32
+			// socket timeout
+			DWORD tv = 10 * 1000; // 10 seconds
+#else
+			// socket timeout
+			struct timeval tv;
+			tv.tv_sec = 10;
+			tv.tv_usec = 0;
+#endif
+			setsockopt(m_msgsock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&tv, sizeof tv);
+			setsockopt(m_msgsock, SOL_SOCKET, SO_SNDTIMEO, (const char*)&tv, sizeof tv);
+
+
+			// either int or ssize_t depending on the platform, so we have to
+			// use a bunch of auto
+			auto receive_length = read_portable(m_msgsock, m_ipc_buffer, MAX_IPC_SIZE);
+			if (receive_length > 0)
+			{
+				// if we already received at least the length then we read it
+				auto end_length = 4;
+				if (receive_length >= end_length)
+				{
+					end_length = FromArray<u32>(m_ipc_buffer, 0);
+					if (end_length > MAX_IPC_SIZE)
+						end_length = MAX_IPC_SIZE;
+				}
+
+				// while we haven't received the entire packet, maybe due to
+				// socket datagram splittage, we continue to read
+				while (receive_length < end_length)
+				{
+					auto tmp_length = read_portable(m_msgsock, &m_ipc_buffer[receive_length], MAX_IPC_SIZE - receive_length);
+
+					// we close the connection if an error happens
+					if (tmp_length <= 0)
+					{
+						receive_length = 0;
+						break;
+					}
+
+					receive_length += tmp_length;
+
+					// if we got at least the final size then update
+					if (end_length == 4 && receive_length >= 4)
+					{
+						end_length = FromArray<u32>(m_ipc_buffer, 0);
+						// we'd like to avoid a client trying to do OOB
+						if (end_length > MAX_IPC_SIZE)
```

Comment:
I think we should respond with an error if they make a request that's too large instead of just ignoring part of it

Otherwise it'll be confusing for the client to debug